c++/src/kj/mutex.c++

```diff
@@ -172,6 +172,65 @@ void Mutex::lock(Exclusivity exclusivity) {
   }
 }
 
+bool Mutex::lockWithTimeout(Exclusivity exclusivity, Duration timeout) {
```

Comment:
Instead of introducing a new method, could we extend the existing `lock()` method with a `Maybe<Duration> timeout` parameter? It looks like most of the (rather complicated) logic is repeated, and the extra logic to handle timeouts shouldn't cost much when the timeout is null.